### PR TITLE
implement async filters

### PR DESCRIPTION
**Pull request by MHBauer**
migration of changes from gerrit. see commit messages for details.

https://gerrit.hyperledger.org/r/c/fabric-chaincode-evm/+/32149

### CODE REVIEW

File: integration/fab3/fab3_test.go [hunk 1]
```diff
@@ -52,6 +61,10 @@ var _ = Describe("Fab3", func() {
 		client        *http.Client
 		SimpleStorage helpers.Contract
 	)
+	core := zapcore.NewCore(zapcore.NewConsoleEncoder(zap.NewDevelopmentEncoderConfig()), zapcore.AddSync(GinkgoWriter), zap.DebugLevel)
```
Comment (swetharepakula) [inline integration/fab3/fab3_test.go]: if you put this in the fab3_suite_test.go in a before each clause it will apply to all the tests including those of the other files.

File: fab3/filters.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
```
Comment (swetharepakula) [inline fab3/filters.go]: I like the new file

File: fab3/filters.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	lastAccessTime time.Time
+}
+
+func (f *logsFilter) LastAccessTime() time.Time {
+	return f.lastAccessTime
+}
+
+func (f *logsFilter) Filter(s *ethService) ([]interface{}, error) {
+	var logs []types.Log
+	err := s.GetLogs(nil, f.gla, &logs)
```
Comment (swetharepakula) [inline fab3/filters.go]: Instead of calling GetLogs every time a filter is asked, i think we should be making use of the event stream that we can get from the sdk. We could cache the events that match a filter and clear that cache per filter when those filter changes are asked. If a filter expires we can remove it then and delete its associated events. 

I think using the event stream will reduce the number of requests we send the peer, especially repeat ones if dapps use multiple filters. It also means we don't need to keep track things like latestBlockSeen. 

I do understand that this implementation reuses a lot of what we have already built, but I think we should optimize on reducing requests to the peer.

**Comment by MHBauer** [comment]
At this point, I'm not inclined to optimize for any use case unless shown some pathological behavior.

We have this implementation, and we should merge it. Then we should follow up with a  new design to use the streams.

File: integration/fab3/fab3_test.go (hunk 1 above)
Comment (MHBauer) [reply integration/fab3/fab3_test.go]: Does the variable show up in the other contexts or do I have to define a global variable to expose it?

Comment (swetharepakula) [reply integration/fab3/fab3_test.go]: you would have to move all four lines, so once you run the ReplaceGlobals it should work on all threads and tests.

**Comment by guoger** [comment]
I emphasize the concern that @swetharepakula had around performance regression, and I think we should definitely address that. Although i think that's beyond the scope of this change - block caching probably deserves a broader review and design, and probably wouldn't impact the code in the PR at all, if done transparently by adding a layer in between.

**Comment by MHBauer** [comment]
It's not a regression, this is currently totally unsupported.

This should have come up during the original design.

If it can indeed be done as a layer between, then we can design that layer separately and put it in later.

**Comment by swetharepakula** [comment]
I wasn't aware that the design was put up anywhere before the the PR. Sorry that I missed that. However, I do think we should continue to have the discussion now, since we did not have the chance before.

Even if the solution is to address certain elements later, I think it is important to come to a consensus now and outline what are the followup tasks to achieve the end design.

**Comment by MHBauer** [comment]
I think the compromise is to merge this and open a new JIRA to have the discussion and design session. There is no underlying structural change that harms any existing functionality.

File: fab3/ethservice_test.go [hunk 4]
```diff
@@ -1322,28 +1322,174 @@ var _ = Describe("Ethservice", func() {
...
-			_, err := strconv.ParseUint(reply, 0, 16)
-			Expect(err).ToNot(HaveOccurred())
-			var valid bool
-			By("using the returned ID to uninstall a filter")
-			Expect(ethservice.UninstallFilter(&http.Request{}, &reply, &valid)).ToNot(HaveOccurred())
-			Expect(valid).To(BeTrue(), "this is the filterID we were given by NewFilter")
-			valid = false // reset to default value
-			Expect(ethservice.UninstallFilter(&http.Request{}, &reply, &valid)).ToNot(HaveOccurred())
-			Expect(valid).To(BeFalse(), "the filter has just now been removed")
+		Context("get filter changes default errors", func() {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: not sure how you want to test getfilterchanegs vs getfilterlogs since they are exactly the same.

File: fab3/ethservice_test.go [hunk 5]
```diff
@@ -1322,28 +1322,174 @@ var _ = Describe("Ethservice", func() {
...
+			It("cannot get a filter when none is installed", func() {
+				filter := types.FilterID{ID: 0}
+				var ret []interface{}
+				Expect(ethservice.GetFilterChanges(&http.Request{}, &filter, &ret)).To(HaveOccurred())
+			})
+			It("cannot get a logs filter when none is installed", func() {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: resolved differently on accident, see above.

File: fab3/ethservice_test.go [hunk 6]
```diff
@@ -1322,28 +1322,174 @@ var _ = Describe("Ethservice", func() {
...
-			Expect(err).ToNot(HaveOccurred())
-			var valid bool
-			By("using the returned ID to uninstall a filter")
-			Expect(ethservice.UninstallFilter(&http.Request{}, &reply, &valid)).ToNot(HaveOccurred())
-			Expect(valid).To(BeTrue(), "this is the filterID we were given by NewFilter")
-			valid = false // reset to default value
-			Expect(ethservice.UninstallFilter(&http.Request{}, &reply, &valid)).ToNot(HaveOccurred())
-			Expect(valid).To(BeFalse(), "the filter has just now been removed")
+		Context("get filter changes default errors", func() {
+			It("cannot get a filter when none is installed", func() {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: same as above

File: Fab3_Instructions.md [hunk 7]
```diff
@@ -368,3 +373,112 @@ curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
...
+}
+```
+
+### eth_getFilterLogs
+
+**Example**
+```console
+curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
+  "jsonrpc":"2.0",
+  "method": "eth_getTransactionCount",
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply Fab3_Instructions.md]: lol

File: fab3/filters.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/hyperledger/fabric-chaincode-evm/fab3/types"
+)
+
+type filterEntry interface {
+	LastAccessTime() time.Time
+	Filter(s *ethService) ([]interface{}, error)
+}
+
+type logsFilter struct {
+	gla            *types.GetLogsArgs
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: Get logs and newfilter support the (nearly) the same input.

I'm surprised it's not exactly the same, and that one doesn't link to the other as that spec is fond of doing.

What do you propose?

File: fab3/filters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: I don't see why. 

The full range is embedded in the logs Args.

**Comment by MHBauer** [comment]
> I think filtering for logs needs to be keeping track of what was the last block seen. Currently it gives all the logs from the beginning of the ledger regardless of when you query it.

The logs args struct contains the block range.

> When a filter expires and it is being removed, currently the methods do not fail. It should fail giving an indication that the the filter has expired.

I think this means in get filter changes.

**Comment by MHBauer** [comment]
rebased as well.

File: integration/fab3/fab3_test.go (hunk 1 above)
Comment (MHBauer) [reply integration/fab3/fab3_test.go]: ya, that's better too cause I can use the zap.S() in any test and have it.

File: fab3/filters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply fab3/filters.go]: Not all logs args will specify a range. Some will be default to latest which means the filter doesn't necessarily end at a certain block?

[the remark this answers is not in the document]
Comment (swetharepakula) [reply fab3/filters.go]: Also say all the blocks in the range have not come in, when you do getFilterChanges, should we only get the logs on the blocks we haven't seen?

**Comment by MHBauer** [comment]
rebase first

File: fab3/filters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: Here's the geth behavior, for log filters:

1. latest is always latest, it updates every time it's called. thus logs made after the filter 
 period do show up.
2. I also updates the earliest base at each call. 

I don't really agree with it, and it's not defined, but it does give us something to target.

```console
# create the filter
➜  ~ curl -v http://localhost:5000 -H "Content-Type: application/json" -XPOST -d '{"jsonrpc":"2.0", "id":37, "params":[ {"address":["0x8Bce8043C94aAe29E9eFb5600669344fdb583457"]}], "method":"eth_newFilter"}'
{"jsonrpc":"2.0","id":37,"result":"0xeeaea1bf4b6c8be083af457b100e2355"}

# don't do anything, get the changes, nothing to get
➜  ~ curl -v http://localhost:5000 -H "Content-Type: application/json" -XPOST -d '{"jsonrpc":"2.0", "id":37, "params":[ "0xeeaea1bf4b6c8be083af457b100e2355" ], "method":"eth_getFilterChanges"}'
{"jsonrpc":"2.0","id":37,"result":[]}

# do one thing, get a log
➜  ~ curl -v http://localhost:5000 -H "Content-Type: application/json" -XPOST -d '{"jsonrpc":"2.0", "id":37, "params":[ "0xeeaea1bf4b6c8be083af457b100e2355" ], "method":"eth_getFilterChanges"}' | jq .
{
  "jsonrpc": "2.0",
  "id": 37,
  "result": [
    {
      "address": "0x8bce8043c94aae29e9efb5600669344fdb583457",
      "topics": [
        "0xd81ec364c58bcc9b49b6c953fc8e1f1c158ee89255bae73029133234a2936aad",
        "0x0000000000000000000000000000000000000000000000000000000000000218",
        "0x00000000000000000000000000000000000000000000000000000000000004d3"
      ],
      "data": "0x",
      "blockNumber": "0xb",
      "transactionHash": "0x01687d3dea50f14d39d5de2fcb52e1c7f4fcab43d6aafa66a18f962103accc26",
      "transactionIndex": "0x0",
      "blockHash": "0x27b86a15a1d9810cdbdd155dcb01078e9ec737be8d71f10577aa1b5b050dec69",
      "logIndex": "0x0",
      "removed": false
    }
  ]
}

# invoke to create a couple logs, get them
➜  ~ curl -v http://localhost:5000 -H "Content-Type: application/json" -XPOST -d '{"jsonrpc":"2.0", "id":37, "params":[ "0xeeaea1bf4b6c8be083af457b100e2355" ], "method":"eth_getFilterChanges"}' | jq .
{
  "jsonrpc": "2.0",
  "id": 37,
  "result": [
    {
      "address": "0x8bce8043c94aae29e9efb5600669344fdb583457",
      "topics": [
        "0xd81ec364c58bcc9b49b6c953fc8e1f1c158ee89255bae73029133234a2936aad",
        "0x00000000000000000000000000000000000000000000000000000000000004d3",
        "0x0000000000000000000000000000000000000000000000000000000000003035"
      ],
      "data": "0x",
      "blockNumber": "0xc",
      "transactionHash": "0xbd10e53f4ddb0863d1783bb020a8bb23c52a90500ae819328b497e7e9d7c33b3",
      "transactionIndex": "0x0",
      "blockHash": "0xaf1fda9c4394d48d3d868b2290f0182e7f5f2a6e94616d1aaf8e45cea2adda34",
      "logIndex": "0x0",
      "removed": false
    },
    {
      "address": "0x8bce8043c94aae29e9efb5600669344fdb583457",
      "topics": [
        "0xd81ec364c58bcc9b49b6c953fc8e1f1c158ee89255bae73029133234a2936aad",
        "0x0000000000000000000000000000000000000000000000000000000000003035",
        "0x000000000000000000000000000000000000000000000000000000000000f389"
      ],
      "data": "0x",
      "blockNumber": "0xd",
      "transactionHash": "0xb4566a08881c501a53b15f3f610a38bc640c94365ec1fe7c879ba9618a25ad11",
      "transactionIndex": "0x0",
      "blockHash": "0xfe7eba70455ee9a674e85dfae54eebdbca596b1a429683a92e23871bd624a972",
      "logIndex": "0x0",
      "removed": false
    }
  ]
}

# don't do any thing, no new logs
➜  ~ curl -v http://localhost:5000 -H "Content-Type: application/json" -XPOST -d '{"jsonrpc":"2.0", "id":37, "params":[ "0xeeaea1bf4b6c8be083af457b100e2355" ], "method":"eth_getFilterChanges"}' | jq .
{
  "jsonrpc": "2.0",
  "id": 37,
  "result": []
}
```

File: fab3/filters.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply fab3/filters.go]: logArgs? They all have have to do with logs

File: fab3/filters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply fab3/filters.go]: I have to verify the behavior on this version of the pr, but my comment was meant to say that the current implementation was returning always from the first block regardless if they those logs had previously been seen. Essentially that we weren't matching geth. I have to look again to see if it does now.

File: fab3/filters.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	return f.lastAccessTime
+}
+
+func (f *logsFilter) Filter(s *ethService) ([]interface{}, error) {
+	var logs []types.Log
+	err := s.GetLogs(nil, f.gla, &logs)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetLogs call failed")
+	}
+	l := make([]interface{}, len(logs))
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: I hate that i have to write this loop everywhere.

File: Fab3_Instructions.md [hunk 10]
```diff
@@ -368,3 +373,112 @@ curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
 
 {"jsonrpc":"2.0","result":"0x0","id":1}
 ```
+
+### eth_NewFilter
+
+**Example**
+```console
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply Fab3_Instructions.md]: removed to match

File: fab3/ethservice.go [hunk 11]
```diff
@@ -549,6 +569,38 @@ func (s *ethService) UninstallFilter(_ *http.Request, filterID *string, removed
...
+		s.filterMapLock.Unlock()
+		return fmt.Errorf("No filter found for id %d", id)
+	}
+	s.filterMapLock.Unlock()
+	// check the filter for validity by expiration time
+	if time.Now().After(f.LastAccessTime().Add(5 * time.Minute)) {
+		// remove the filter if not valid and return error
+		// delegate to existing function
+		var removed bool
+		return s.UninstallFilter(nil, filterID, &removed)
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice.go]: that's basically what the comment above says.

I don't think there's any diagnostic information to be gained by erroring on timeout vs returning false that the filter was not present.

I also don't think there's much information in the single bit of true/false we're giving back for "we found it and removed it" vs "we didn't find it" vs "we found it, and it was timed out".

I can't very well conceive of anything harmful in the current behavior.

File: Fab3_Instructions.md [hunk 12]
```diff
@@ -26,6 +26,11 @@ Fab3 currently supports:
 - [eth_getTransactionReceipt](#eth_getTransactionReceipt)
 - [eth_getLogs](#eth_getLogs)
 - [eth_getTransactionCount](#eth_getTransactionCount)
+- [eth_newFilter](#eth_newFilter)
+- [eth_newBlockFilter](#eth_newBlockFilter)
+- [eth_yninstallFilter](#eth_uninstallFilter)
```
Comment (MHBauer) [inline Fab3_Instructions.md]: messed this one up

File: fab3/ethservice.go [hunk 13]
```diff
@@ -531,21 +535,36 @@ func (s *ethService) NewFilter(_ *http.Request, filter *types.GetLogsArgs, resul
...
+	id := filterID.ID
 	s.filterMapLock.Lock()
 	defer s.filterMapLock.Unlock()
 
+	// do we care about timing out filters while removing them?
+	// i.e. turning the successful delete if present into an error
+	// because it timed out?
 	if _, ok := s.filterMap[id]; ok {
 		delete(s.filterMap, id)
 		*removed = true
```
Comment (MHBauer) [inline fab3/ethservice.go]: that's basically what the comment above says.

I don't think there's any diagnostic information to be gained by erroring on timeout vs returning false that the filter was not present.

I also don't think there's much information in the single bit of true/false we're giving back for "we found it and removed it" vs "we didn't find it" vs "we found it, and it was timed out".

I can't very well conceive of anything harmful in the current behavior.

File: fab3/ethservice.go [hunk 14]
```diff
@@ -80,7 +81,11 @@ type EthService interface {
 	GetTransactionCount(r *http.Request, _ *interface{}, reply *string) error
 	GetLogs(*http.Request, *types.GetLogsArgs, *[]types.Log) error
 	NewFilter(*http.Request, *types.GetLogsArgs, *string) error
-	UninstallFilter(*http.Request, *string, *bool) error
+	NewBlockFilter(*http.Request, *interface{}, *string) error
+	// NewPendingTransactionFilter (*http.Request, *types.GetLogsArgs, *string) error
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice.go]: 410 Gone

**Comment by MHBauer** [comment]
no idea what's up with the unit test failure in ci build. 

```
cd fab3 && GO111MODULE=on ginkgo -p -randomizeAllSpecs -randomizeSuites -requireSuite -noColor -keepGoing -race -r
Failed to compile types:

# golang.org/x/sys/unix
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:7: const initializer _CPU_SETSIZE / _NCPUBITS is not a constant
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:20: undefined: _CPU_SETSIZE
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:35: undefined: _NCPUBITS
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:17:25: undefined: cpuMask
note: module requires Go 1.12Failed to compile fab3:

go: finding github.com/hyperledger/fabric-chaincode-evm/evmcc v0.0.0
# golang.org/x/sys/unix
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:7: const initializer _CPU_SETSIZE / _NCPUBITS is not a constant
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:20: undefined: _CPU_SETSIZE
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:14:35: undefined: _NCPUBITS
../../../../../pkg/mod/golang.org/x/sys@v0.0.0-20191020212454-3e7259c5e7c2/unix/affinity_linux.go:17:25: undefined: cpuMask
note: module requires Go 1.12
There were failures detected in the following suites:
	types ./types
	 fab3 .

```

File: fab3/filters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: Yes, exactly. There's nothing in the spec that mandates that behavior. I hadn't run geth before now, so the geth behavior was not a target.

File: fab3/filters.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/hyperledger/fabric-chaincode-evm/fab3/types"
+)
+
+type filterEntry interface {
+	LastAccessTime() time.Time
+	Filter(s *ethService) ([]interface{}, error)
+}
+
+type logsFilter struct {
+	logArgs        *types.GetLogsArgs
```
Comment (swetharepakula) [inline fab3/filters.go]: Because of how rebasing and I guess github works with comments, we lost the discussion on the GetLogsArgs naming. I am thinking something like `FilterArgs` or `LogFilterArgs` as the name would be good. But since this is a rather large PR already, either we should do it a different PR or at least in a separate commit.

File: Fab3_Instructions.md [hunk 16]
```diff
@@ -369,3 +374,124 @@ curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
...
+**Example**
+```
+curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
+  "jsonrpc":"2.0",
+  "method": "eth_getFilterChanges",
+  "id": 6129484611666146000,
+  "params":["0x2"]
+}'
+
+{
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply Fab3_Instructions.md]: Not sure how to differentiate it. Could do one of each and have comments at the top of each block.

File: fab3/filters.go (hunk 15 above)
Comment (MHBauer) [reply fab3/filters.go]: previous suggestion was logargs https://github.com/hyperledger/fabric-chaincode-evm/pull/11#discussion_r364001268

File: fab3/types/types.go [hunk 17]
```diff
@@ -48,12 +49,29 @@ type GetLogsArgs struct {
 	BlockHash string        `json:"blockHash,omitempty"`
 }
 
+type FilterID struct {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/types/types.go]: It's not only a number, it's a specific kind of number. I am using the typesystem to help me out. 

I'm not sure how golang deals with type aliases. Need to check if we can derive or extend from one struct, and still not have them substitute.

File: fab3/filters.go (hunk 15 above)
Comment (swetharepakula) [reply fab3/filters.go]: Ohh, I see now. I miss gerrit, it would been able to highlight the right word. My comment was about the type name not the field name. I think the field name `logArgs` is fine. I do prefer that over `gla`

File: fab3/ethservice_test.go [hunk 18]
```diff
@@ -1340,28 +1340,174 @@ var _ = Describe("Ethservice", func() {
...
+					id, err := strconv.ParseUint(reply, 0, 16)
+					Expect(err).ToNot(HaveOccurred())
+					var ret []interface{}
+
+					var filter types.FilterID
+					By("put some blocks into the chain")
+					mockLedgerClient.QueryInfoReturns(&fab.BlockchainInfoResponse{BCI: &common.BlockchainInfo{Height: 3}}, nil)
+					sampleBlock1 := GetSampleBlock(1)
+					sampleBlock2 := GetSampleBlock(2)
+					qbs := func(b uint64, _ ...ledger.RequestOption) (*common.Block, error) {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: makes sense. what would be write out the variable? queryblockstub?

File: fab3/ethservice_test.go [hunk 19]
```diff
@@ -1340,28 +1340,174 @@ var _ = Describe("Ethservice", func() {
...
+					var reply string
+					Expect(ethservice.NewBlockFilter(&http.Request{}, nil, &reply)).ToNot(HaveOccurred())
+					id, err := strconv.ParseUint(reply, 0, 16)
+					Expect(err).ToNot(HaveOccurred())
+					filter := types.FilterID{ID: id}
+					var ret []interface{}
+					mockLedgerClient.QueryInfoReturns(&fab.BlockchainInfoResponse{BCI: &common.BlockchainInfo{Height: 2}}, nil)
+					mockLedgerClient.QueryBlockReturns(nil, fmt.Errorf("ledger can't get blocks"))
+					Expect(ethservice.GetFilterChanges(&http.Request{}, &filter, &ret)).To(HaveOccurred())
+				})
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: nothing for newblockfilters?

File: fab3/ethservice_test.go [hunk 20]
```diff
@@ -1340,28 +1340,174 @@ var _ = Describe("Ethservice", func() {
...
+	Describe("Filtering", func() {
+		Context("NewFilter & UninstallFilter", func() {
+			It("fails to find filters to uninstall when none are installed", func() {
+				id := rand.Uint64()
+				filter := types.FilterID{ID: id}
+				var valid bool
+				Expect(ethservice.UninstallFilter(&http.Request{}, &filter, &valid)).ToNot(HaveOccurred())
+				Expect(valid).To(BeFalse())
+			})
+			It("have a consistent Filter ID between invocations of NewFilter and UninstallFilter", func() {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/ethservice_test.go]: I can split this into two tests if it's confusing, but I'm not testing the uninstallfilters that don't exist.

I'm making sure the new filter that was created is visiblet o the uninstallfilter with the exact Id that newfilter returned.

File: fab3/filters.go [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+func (f *logsFilter) LastAccessTime() time.Time {
+	return f.lastAccessTime
+}
+
+func (f *logsFilter) Filter(s *ethService) ([]interface{}, error) {
+	var logs []types.Log
+	err := s.GetLogs(nil, f.logArgs, &logs)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetLogs call failed")
+	}
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: ah, "since last poll", okay. I failed reading comprehension that day.

[the remark this answers is not in the document]
Comment (MHBauer) [reply fab3/filters.go]: and filterlogs is all, how terrible an API.

File: Fab3_Instructions.md [hunk 22]
```diff
@@ -369,3 +374,124 @@ curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
...
+  "params":["0x2"]
+}'
+
+{"jsonrpc":"2.0","result":true,"id":1}
+```
+
+### eth_getFilterChanges
+`eth_getFilterChanges` takes a filter identifier and returns the output
+associated with the filter. For new block filters, that is an array of block
+hashes. For log filters, it is the log entries as if from `eth_getLogs`.
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply Fab3_Instructions.md]: should be done in #27

File: Fab3_Instructions.md [hunk 23]
```diff
@@ -369,3 +374,124 @@ curl http://127.0.0.1:5000 -X POST -H "Content-Type:application/json" -d '{
 
 {"jsonrpc":"2.0","result":"0x0","id":1}
 ```
+
+### eth_newFilter
+`eth_newFilter` takes the same arguments as `eth_getLogs`. It returns an
+identifier to collect the log entries. The log filter is not run until
+`eth_getFilterChanges` is called.
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply Fab3_Instructions.md]: should be done in #27

File: integration/fab3/fab3_test.go [hunk 24]
```diff
@@ -378,31 +393,49 @@ var _ = Describe("Fab3", func() {
...
-		rBody, err = ioutil.ReadAll(resp.Body)
+		By("successfully uninstalling the installed filter")
+		rBody = sendRPCRequest(client, "eth_uninstallFilter", proxyAddress, 38, filterID)
+
+		err = json.Unmarshal(rBody, &uninstallBody)
 		Expect(err).ToNot(HaveOccurred())
-		fmt.Fprintln(GinkgoWriter, string(rBody))
+		Expect(uninstallBody.Result).To(BeTrue())
+
+		By("lacking presence of the installed filter")
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply integration/fab3/fab3_test.go]: done in #27

**Comment by swetharepakula** [comment]
Can this be closed in favor of #27 ?

**Comment by swetharepakula** [comment]
Closing this PR, since the changes have been split into smaller PRs and last one left is #27.
